### ISSUE TITLE
CMD-Q won't close napari when "install plugins..." window is open

### ISSUE THREAD

**Issue by neuromusic** [napari/napari#2490]
## 🐛 Bug

CMD-Q won't close napari when "install plugins..." window is open

## To Reproduce

1. Open the bundled app
2. Open the plugin installation window
3. Press "CMD-Q" (on mac)
4. Nothing happens

## Expected behavior

Expected napari to close as it does when the plugin installation window is NOT open

## Environment

napari: 0.4.6
Platform: macOS-10.15.7-x86_64-i386-64bit
System: MacOS 10.15.7
Python: 3.8.7 (default, Jan 2 2021, 04:16:43) [Clang 11.0.0 (clang-1100.0.33.17)]
Qt: 5.15.2
PySide2: 5.15.2
NumPy: 1.19.3
SciPy: 1.6.1
Dask: 2021.03.0
VisPy: 0.6.6

OpenGL:
- GL version: 2.1 ATI-3.10.19
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1792x1120, scale 2.0

Plugins:
- console: 0.0.3
- dv_reader: 0.2.5
- dzi-zarr: 0.1.1
- napari-czifile2: 0.1.1
- napari-demo: 0.0.1
- napari-em-reader: 0.1.0
- napari-hdf5-labels-io: 0.2.dev1
- napari-imc: 0.5.1
- napari-nikon-nd2: 0.1.3
- napping: 0.1.6
- ndtiffs: 0.1.1
- svg: 0.1.4

**Comment by sofroniewn** [comment]
@goanpeca @ppwadhwa maybe one of you might be able to take a look at this sometime. Thanks!!

**Comment by goanpeca** [comment]
Will do!

**Comment by goanpeca** [comment]
It seems we either:
- Make it a [non modal dialog](https://lists.qt-project.org/pipermail/qt-interest-old/2010-October/027911.html)?
- Have some way of adding/rebinding application shortcuts to any dialog opened as modal.

@napari/core-devs ?

**Comment by tlambert03** [comment]
yeah, this really is broader issue than the plugin dialog, and comes down to whether we _ever_ want to have modal dialogs, and if so, whether we want to override Qt's default behavior of having to exit the modal before being able to use global key bindings.  I feel like I've seen programs do both (that is, some programs _do_ have windows that must be attended to before you can quit the entire program), but I don't know whether there is a UX "standard" here.  I think we will probably want modal dialogs somewhere at some point in napari, so I think it comes down to a decision between your second option (the rebinding option), or deciding that we're ok with modals being blocking...

if we decide to make everything non-modal, I suspect we could quite easily turn into a many-windowed application... and avoiding that has been a core napari "principle" from early on.

**Comment by goanpeca** [comment]
On this comment:

>if we decide to make everything non-modal, I suspect we could quite easily turn into a many-windowed application... and avoiding that has been a core napari "principle" from early on.

That goes a bit against how applications "should" work on Mac (the OS all of the napari devs seem to be using 🙃 ?) where modals are to be used sparingly (as mentioned on the HIGs)

Just my 2 cents :)

**Comment by tlambert03** [comment]
> where modals are to be used sparingly

I wasn't implying that modals are the (or at least the only) solution to window explosion... just that it's something we'd very much like to avoid

**Comment by goanpeca** [comment]
>I wasn't implying that modals are the (or at least the only) solution to window explosion... just that it's something we'd very much like to avoid

Did not mean to imply that for windows in general, but for the specifics of dialogs in napari (which also came with the [preferences dialog](https://github.com/napari/napari/pull/2457)) I am not aware of any workaround besides making them all modal.

**Comment by sofroniewn** [comment]
If I do `cmd+Q` from this window (or file quit) then it actually works

![image](https://user-images.githubusercontent.com/6531703/113196032-dc213900-9217-11eb-946c-7b910efbedcb.png)

Maybe this was changed for `0.4.7` or maybe the issue was no keybindings working unless focus changes (i still get that without launching for pythonx), or maybe I'm missing something here.

I get the concerns of the larger convo, and am not quite sure what to recommend ....

------------------------------------------------------------------------------------------------------------

> Side note, do we have some numbers on usage of napari per OS?

Here are some pypi metrics and napari.org website visit stats by OS, it's definitely mixed

![image](https://user-images.githubusercontent.com/6531703/113195367-0cb4a300-9217-11eb-9697-84f6f016b3e1.png)

![image](https://user-images.githubusercontent.com/6531703/113195267-ed1d7a80-9216-11eb-9901-1d8c5ac6ac13.png)

> It might be worth considering at some point to have QA testing for Windows and linux. While we rely on Qt/PyQt, some decisions we might take (or have taken in the past) may be going against guidelines for different OSs.

Yes agreed!!!

**Comment by neuromusic** [comment]
super interesting re: the more OS-level enforcement of modal dialog behavior. thank you for providing this context @goanpeca & @tlambert03 !!

> If I do cmd+Q from this window (or file quit) then it actually works

I see "python" in the menu bar.... is that from the standalone/bundled app? or from napari initialized from Python?

**Comment by sofroniewn** [comment]
> @sofroniewn : I see "python" in the menu bar.... is that from the standalone/bundled app? or from napari initialized from Python?

mine was launching via a script not using pythonx

**Comment by goanpeca** [comment]
Ok so on current `master` it is not working for me :-|, so there is definitely something going, but I believe it is enforced by Qt by default.

I would suggest some of the following, leveraging the work that @Carreau is doing for the actions manager. We can have the base dialog on Napari (we have one right?) take the actions defined in  the Applitcation context and apply them to the dialog on `__init__`, something like

```python
class NapariBaseDialog(QDialog):

    def __init__(self):
        super().__init__()
        self.addActions(actions_manager.application_actions())
```

Thoughts @sofroniewn @tlambert03 ?

**Comment by psobolewskiPhD** [comment]
Just to re-fresh this, with the new Plugin Manager (main) on macOS neither Command-Q nor Command-W do anything (trigger quit of napari app nor close the manager).
I think it's reasonable for Command-Q to be blocked, but Command-W should work, assuming any install/uninstall/update as completed.

**Comment by Czaki** [comment]
Few technical detils:

1. Currently, the closing window/app is implemented on the level of Napari main window.
2. Plugin dialog is QtDialog working in [modal mode](https://doc.qt.io/qt-6/qdialog.html#modal-dialogs)
3. We use Main Window to pass shortcuts to our internal mechanism in QtViewer https://github.com/napari/napari/blob/9da6d289735e4c7d3abbcd111f03f5c61891d7ff/napari/_qt/qt_viewer.py#L847

So if we would like to close Plugin Dialog with shortcuts, it must be implemented in Plugin Dialog. 

If there is a need for more details, I can try to provide it. 
But at this moment, I suggest transferring it to https://github.com/napari/napari-plugin-manager/
